### ISSUE TITLE
Container state seems to be preserved in the engine

### ISSUE THREAD

**Issue by srfrnk** [moby/moby#37837]
While researching a bug with Kafka and node-kafka which is unrelated to docker - I've stumbled upon a problem with docker engine.

After running with Kafka 2.0.0 you cannot go back to use the working Kafka 1.1.0 version without restarting the Doccker Service.

Even removing all container, images and volumes and then doing a `docker system prune -a --volumes` doesn't help.

The only thing that does is running `service docker restart` - no need to 'prune' - just restart.

I would expect state to be restored w/o even pruning - just killing the container should be enough.

The complete source code and environment [can be found here](https://github.com/srfrnk/kafka-test)
The kafka/node-kafka bug [can be found here](https://github.com/SOHU-Co/kafka-node/issues/1073)

**Comment by cpuguy83** [comment]
Can you explain what you mean? What are you doing, what did you expect to happen, etc? We have an issue template for this reason.

**Comment by srfrnk** [comment]
1) Clone [this](https://github.com/srfrnk/kafka-test)
1) In terminal run ```make run``` - Notice it's running OK
1) Edit `docker-compose.yml` and change:
```
kafka:
    image: wurstmeister/kafka:1.1.0
    # image: wurstmeister/kafka:latest
```

To:
```
kafka:
    image: wurstmeister/kafka:1.1.0
    # image: wurstmeister/kafka:latest
```
1) In terminal run ```make run``` - Notice it's NOT running OK
1) Edit `docker-compose.yml` and change:
```
kafka:
    # image: wurstmeister/kafka:1.1.0
    image: wurstmeister/kafka:latest
```

To:
```
kafka:
    # image: wurstmeister/kafka:1.1.0
    image: wurstmeister/kafka:latest
```

1) In terminal run ```make run``` - Notice it's NOT running OK

Expected:
The Kafka:1.1.0 image should work the second time.
Each execution of a docker container should be pristine (i.e. not affected by previous containers run)

**Comment by cpuguy83** [comment]
I'm not exactly following.

What is the state of containers in docker first, second, third, fourth, what do you expect?
What version of docker are you using?

**Comment by srfrnk** [comment]
```bash
$ docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:15 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

What do you mean by 'state of containers...'?
The expected behavior I think is that each container be started with a 'clean slate' - not inherits any information/state/behavior from previously run containers - am I wrong?
One container run should not effect another container run (unless they are linked by network/mounts) right?

This scenario demonstrates a use case in which two container being run sequentially have effect on one another where such shouldn't exist.

Is that more clear?

**Comment by cpuguy83** [comment]
> What do you mean by 'state of containers...'?
As in what does Docker say about them in "docker ps" and "docker inspect".

If you are running this from compose, then multiple runs generally do share state as compose is more project based.

**Comment by srfrnk** [comment]
So is it an expected behavior for docker compose to run two different containers with two different images and have what I suspect to be shared TCP connections? Is that not an issue with the isolation of the network layer between containers?
And to emphasize - the docker compose is being killed between runs - i.e. `docker-compose down`

**Comment by srfrnk** [comment]
I understand your concern.
If I could find a way to simplify the use case I would.

I will try to explain what happened and how I came up with this use case.
Basically what is being done here is running a Kafka cluster using docker compose.
To do so it create a Zookeeper container from image: [`zookeeper`](https://hub.docker.com/_/zookeeper/)
Then it creates a Kafka container which connects to the ZK container and uses the image from [`wurstmeister/kafka`](https://hub.docker.com/r/wurstmeister/kafka/)
This constitutes the most minimal Kafka cluster one can create.

To use the Kafka cluster a NodeJS container is also created.
This container uses [`node'](https://hub.docker.com/_/node/) image.
It imports the npm `kafka-node` package to connect to the Kafka cluster.

What happened to me is I stumbled upon a known bug between `kafka-node` and latest Kafka.
The package was working with Kafka v1.1.0 but not the latest.

To keep using it I killed the entire docker-compose application (ZK cluster, Kafka cluser etc...) and started a new one only with v1.1.0 for the kafka image. However to my surprise it kept not working -  the bug persisted.

After many attempts I finally realized it was docker that created the issue.

If I reset my machine then started the 1.1.0 Kafka image - all worked.
However running the latest Kafka - will make any forther attempt fail until the machine is restarted or the docker engine service restarted.

This is why I suspect some kind of state is being preserved by the engine which should not.

The bug that `kafka-node` has with latest Kafka has something to do with TCP connections not being properly closed.

This is why I suspect that the improperly maintained state has something to do with TCP connections.

This is just my educated guess of course since I cannot find a more minimal use case to demonstrate the problem.

Hope that helps a bit.
